neuralpp/experiments/experimental_inference/graph_computation.py

```diff
@@ -0,0 +1,110 @@
+from neuralpp.util.util import argmax
+
+
+class PartialTreeComputation:
+    """
+    Defines a computation evaluated on partial_tree, which can be updated
+    as the partial tree gains edges from the full tree.
+
+    Requires that partial_tree is a subtree of full_tree. (TODO: Add validation for this)
+
+    * Get syntax is used to retrieve cached values of the computation.
+    * Call syntax will evaluate the function over the partial tree, updating the result_dict. (TODO: do we need this?
+      in most cases it's only used during initialization, as long as results are propagated when changed.)
+    """
+
+    def __init__(self, partial_tree, full_tree):
+        self.partial_tree = partial_tree
+        self.full_tree = full_tree
+        self.result_dict = {}
+
+    def __getitem__(self, item):
+        return self.result_dict[id(item)]
+
+    def __setitem__(self, key, value):
+        self.result_dict[id(key)] = value
+
+    def __call__(self, item):
+        self.compute_result_dict(item)
+        return self.result_dict[id(item)]
+
+    def compute_result_dict(self, node):
+        raise NotImplementedError()
+
+    def update_value(self, target_node, changed_child, child_value):
+        raise NotImplementedError()
+
+    def recompute_and_propagate_result_to_parents(self, node):
+        self.compute_result_dict(node)
+        node_result = self[node]
+        target_node = self.partial_tree.parent(node)
+        while target_node is not None:
+            self.update_value(target_node, node, node_result)
+            node = target_node
+            target_node = self.partial_tree.parent(target_node)
+
+
+class ExpansionValueComputation(PartialTreeComputation):
```

Comment:
I think we need to rethink the design of this expansion tree; it seems to be dong several different things at once. Let's talk about it in our meeting today.